### PR TITLE
Gen 2 Random Battles: General update and improvements

### PR DESCRIPTION
**Pull request by xCrystal**
## Introduction

This is a complete overhaul of Gen 2 Random Battles. It only covers fully evolved Pokemon and Scyther. I did update Gen 1 Random Battles not so long ago, and then, I decided to move on to Gen 2, given that it was in a much worse state. Due to the higher depth of this generation however, it was also going to be a lot more work.

As of know, moveset generation in Gen 2 RB is really awful. Most movepools consist on a bunch of viable moves thrown together (pressumably copied directly from the Smogon GSC OU analyses) which include a lot filler or niche moves. When four moves are picked up at random, this can lead to very bad results. My personal favorite is Moltres with Sunny Day / Hidden Power Water / Sleep Talk / Reflect. Granted the Gen 2 scripts contain some basic sanity rules but they still don't prevent things like Sleep Talk without Rest everywhere or many Pokemon showing up without their main STAB move.

## Movesets

While improving RB to a decent state wouldn't be that much work, I still decided to take it to the next level and completely rework the RB scripts and movesets. I tried to find the optimal combination of movesets considering that this in unlike a metagame like OU, where Pokemon of any tier can show up and their levels are balanced acoring to their tier. Therefore, moves like Spikes, Roar and Heal Bell won't as prevalent, and the metagame will be generally faster paced. 

That being said, this was my approach:

- Pokemon have between one and five possible movesets, each of them having a chance between 1/16 and 16/16 assigned to them. This allows to have niche movesets (that are still viable) along with the main or more solid movesets, providing variety and unpredictability.
- Each moveset consists of 0-4 base moves and of 0-4 filler moves, for a total of 4 moves. Base moves are the mandatory moves that are always included, like the main STAB move or various moveset-defining moves. Filler moves are arrays of moves where one move is picked at random. 
- The item is randomly picked from an array of possible items (usually it's just a single "leftovers").

As a quick example, this is how Lapras data looks like:

		randomSet1: { // RestTalk
			chance: 8,
			item: ["leftovers"],
			baseMove1: "rest", baseMove2: "sleeptalk", baseMove3: "icebeam",
			fillerMoves1: ["thunder", "thunderbolt"],
		},
		randomSet2: { // Attacker
			chance: 16,
			item: ["leftovers"],
			baseMove1: "icebeam",
			fillerMoves1: ["thunderbolt", "thunderbolt", "thunder"],
			fillerMoves2: ["hydropump", "surf"],
			fillerMoves3: ["rest", "sing", "confuseray", "reflect"],
		},

I should probably add, I left the move Attract out of the picture. Whether players take care of adjusting their Pokemon DVs to make a Pokemon female or not has always been up to them, plus a female Pokemon tends to reveal that it has no physical moves or Hidden Power. In the end, it felt like manipulating the effectiveness of some Pokemon a bit too much, so nothing was given Attract or had their DVs adjusted for gender purposes. I also left sleep trapping out due to being sometimes banned (and admittedly, because it would be a bit broken on some Pokemon). And of course, the likes of Double Team and OHKO moves are also not present.

## Scripts
Due to the depth in the moveset structures, there's no need to work on rules to prevent redundancy. However, I did work on various filters to encourage variety and ensure a well-rounded team. Here's a summary:

### Pokemon filters
- 1 Uber at most
- Not more than 2 Pokemon of the same tier (OU/BL/UU/NU)
- The same type not more than twice
- 1/3 chance to discard single-type Pokemon if that type already exists in the team
- 2/3 chance to discard double-type Pokemon if both types already exist in the team
- Not more than -2 overall weakness to the same type (e.g. you can draw a third Pokemon weak to a type if you already have one resisting or immune to it)

While the combination of all filters may seem very restrictive, there are a whole lot of Pokemon to ensure that we never run out. There are still, however, failsafes to ensure that Pokemon aren't discarded when the amount of slots to fill is equal to the Pokemon left. Regardless, during my tests, out of the pool of 132 Pokemon, the lowest Pokemon remaining I ever saw before filling up the team was 92, and it would very rarely drop below 115.

### Move filters
- The following moves not more than once: Reflect, Light Screen, Rapid Spin, Spikes, Belly Drum, Haze, Heal Bell, Thief, phazing move (Roar/Whirlwind)
- The following moves not more than twice: Sleep Talk, sleeping move (Sleep Powder, Hypnosis...)
- All the above moves except Belly Drum, Haze and Thief are slightly encouraged (compared to their overall presence in the movesets) to exist at least once in the team

## Misc
I assumed this would take a while to merge (there are also conficts) so I purposely made the PR with a few things here and there to finish.

I made some rushed up function to test to ensure that any mistakes and typos are caught, which needs to be deleted now. There are also some random console.log's to delete, marked with a TODO tag. **_cleaned up already_**

The main thing that is left to do know is adjusting the DVs of a Pokemon to its Hidden Power, given that they all get 15 all by default. It shouldn't be too difficult assuming that modifying the DVs will lead to the correct stats, like it does in OU. _**done already**_

I found there's another problem, too. Thick Club and Metal Powder aren't providing boosts to Marowak and Ditto, respectively. They clearly work fine in OU, but for some reason, they do not work in RB. Someone more familiar with the engine than me may be able to find out why (I guess worst case scenario would be adjusting Marowak's moveset for Leftovers but I hope I don't have to resort to that). _**fixed already**_

I had also planned additional Pokemon generation filters to ensure a minimum of offense-oriented Pokemon and/or an overall minimum of damaging moves, in order to prevent overly stally teams. I'm a bit burned out with this though, so I probably won't be doing it. Overall it tends to be faster paced than OU though, and I did try to make sure to adjust the odds of defensive movesets so that they aren't too high overall (granted, there are Pokemon only capable of running defensive movesets, and you're not going to corrupt them for the same of speeding up the whole thing). _**not going to work on this, at least not on this PR**_

Anyway, I guess that's all. I opened this thread in Smogon when I started working on it (http://www.smogon.com/forums/threads/improving-gen-2-random-battles-on-ps.3597279/), but I've pretty much done it all by myself anyway. The main goal is to have Gen 2 RB in the ladder. We still have Gen 2 RB tournaments in the Ruins of Alph room quite frequently and it hurts to see how bad the movesets are!

1c77170 fixes the merge conflict (I had accidentally deleted Pichu's event moves)

**Ready for merging**

### CODE REVIEW

**Comment by Zarel** [comment]
Tell me when this is mergeable. Also, @DisasterAreaPS and @TheImmortal, you guys might want to review.

**Comment by xCrystal** [comment]
Yeah, I'll take care of the remaining things tomorrow and clean everything up. I'll post again here whenever it's ready.

**Comment by TheImmortal** [comment]
I don't know Gen 2 so I can't really review the specific sets. As for the new structure, it's fine. Older gens are probably not going to get both Battle Factory and Random Battle. I welcome @xCrystal's overhaul and am happy to see someone running gen 1/2 Random Battle.

**Comment by xCrystal** [comment]
Except for the Thick Club (and Metal Powder) bug, this should be finished and ready to be merged whenever.

**Comment by xCrystal** [comment]
I see `Tools` has been renamed as `Dex`, is that right?

**Comment by xCrystal** [comment]
All good now!

**Comment by Zarel** [comment]
Yes, `Tools` has been renamed `Dex` and has a slightly different API now.

**Comment by Zarel** [comment]
You should be using `this` in scripts, anyway...

File: mods/gen2/scripts.js [hunk 1]
```diff
@@ -535,311 +535,294 @@ exports.BattleScripts = {
...
+			} else if (types.length === 2) {
+				if (typeCount[types[0]] > 1 || typeCount[types[1]] > 1) skip = true;
+				if (typeCount[types[0]] && typeCount[types[1]] && this.random(3) > 0) skip = true;
+			}
+
+			// Ensure the weakness-resistance balance is 2 points or lower for all types,
+			// but ensure no more than 3 pokemon weak to the same regardless.
+			let weaknesses = [];
+			for (let type in weaknessCount) {
+				let weak = Dex.getImmunity(type, template) && Dex.getEffectiveness(type, template) > 0;
```
Comment (Zarel) [inline mods/gen2/scripts.js]: i.e. this should be `this.getImmunity`, `this.getEffectiveness`

**Comment by Zarel** [comment]
The reason is because `Dex` is a gen 7 dex, and `this` is a dex for the current generation.

**Comment by xCrystal** [comment]
How do `getImmunity` and `getEffectiveness` get bound to `this`, if I may ask? My basic knowledge of JavaScript/Node is that if you export/import something you reference it with the module name, whereas with `this`you can only reference things in the current scope.

**Comment by Zarel** [comment]
In a PS Script context, (i.e. basically every function in `data/`), `this` refers to the `Battle` object running the script.

`Battle` is a subclass of `Dex` (technically `ModdedDex`), so battles have access to all the functions in `Dex`.

**Comment by DisasterAreaPS** [comment]
Had a chat with Crystal_ as well as Enigami and xJoelituh in our RBY Development Discord DM group and it all sounds sensible to me. Haven't reviewed the sets but I'm not as much of a GSC expert as I am in RBY or ADV by far; I trust Crystal_ to have done a good job.

File: mods/gen2/scripts.js [hunk 2]
```diff
@@ -51,9 +51,9 @@ exports.BattleScripts = {
 			}
 
 			// Treat here the items.
-			if ((this.species in {'Cubone':1, 'Marowak':1} && this.item === 'thickclub' && statName === 'atk') || (this.species === 'Pikachu' && this.item === 'lightball' && statName === 'spa')) {
+			if ((this.name in {'Cubone':1, 'Marowak':1} && this.item === 'thickclub' && statName === 'atk') || (this.name === 'Pikachu' && this.item === 'lightball' && statName === 'spa')) {
 				stat *= 2;
-			} else if (this.species === 'Ditto' && this.item === 'metalpowder' && statName in {'def':1, 'spd':1}) {
+			} else if (this.name === 'Ditto' && this.item === 'metalpowder' && statName in {'def':1, 'spd':1}) {
```
Comment (urkerab) [inline mods/gen2/scripts.js]: So if my Dragonite is named Cubone and I give it a Thick Club will that double its attack?

Comment (xCrystal) [inline mods/gen2/scripts.js]: Whoops, good catch. Luckily, PS doesn't let you nickname Pokemon exactly like the name of another Pokemon, but due to this, nicknamed Marowak and the like won't get their boosts. Damn, I feel dumb now for not just renaming moveset.name to moveset.species...

https://github.com/Zarel/Pokemon-Showdown/pull/3658

**Pull request merged**